### PR TITLE
Try using slobrok.internal.fetchLocalView also

### PR DESCRIPTION
**Pull request by arnej27959**
* add a MapView class holding a mirror of remote slobrok
* perform incremental fetch to fill the MapView when possible.

I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

@havardpe please review

### CODE REVIEW

File: slobrok/src/vespa/slobrok/server/map_view.cpp [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+namespace slobrok {
+
+MapView::MapView()
+  : _map(),
+    _currGen(0),
+    _lock()
+{}
+
+MapView::~MapView() {
+    clear();
```
Comment (havardpe) [inline slobrok/src/vespa/slobrok/server/map_view.cpp:16]: This implies that we are ok with adding listeners that are never removed. Should we be ok with that?

File: slobrok/src/vespa/slobrok/server/map_view.cpp [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+}
+
+void MapView::apply(const MapDiff &diff) {
+    std::lock_guard guard(_lock);
+    LOG(debug, "Applying diff from gen %u", diff.fromGen.getAsInt());
```
Comment (havardpe) [inline slobrok/src/vespa/slobrok/server/map_view.cpp:21]: consider adding assert that from gen matches (map vs diff) to indicate that this needs to be handled externally.

File: slobrok/src/vespa/slobrok/server/map_view.h [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+#include <mutex>
+#include <set>
+
+namespace slobrok {
+
+/**
+ * @class MapView
+ * @brief Holds a name->spec map which can be incrementally updated
+ **/
+class MapView : public MapSource
```
Comment (havardpe) [inline slobrok/src/vespa/slobrok/server/map_view.h:19]: MapView sounds a bit generic. Other options could be RemoteMapSource (we do it for the events) or ServiceMapMirror or similar (to match it with its partner; ServiceMapHistory)

File: slobrok/src/vespa/slobrok/server/remote_slobrok.h [hunk 4]
```diff
@@ -67,9 +71,11 @@ class RemoteSlobrok: public IRpcServerManager,
     bool isConnected() const { return (_remote != nullptr); }
     void tryConnect();
     void maybePushMine();
+    void maybeStartFetch();
     void invokeAsync(FRT_RPCRequest *req, double timeout, FRT_IRequestWait *rwaiter);
     const std::string & getName() const { return _rpcserver.getName(); }
     const std::string & getSpec() const { return _rpcserver.getSpec(); }
+    const MapView &remoteView() const { return _mapView; }
```
Comment (havardpe) [inline slobrok/src/vespa/slobrok/server/remote_slobrok.h:78]: we should probably return MapSource here, which raises the question about mutable event listener lists...

**Review by havardpe — APPROVED** [review]
looks reasonable, but also touches on code that really does not want this to run with multiple threads (which makes the locking only half-way useful).

File: slobrok/src/vespa/slobrok/server/map_view.cpp (hunk 1 above)
Comment (arnej27959) [reply slobrok/src/vespa/slobrok/server/map_view.cpp]: > This implies that we are ok with adding listeners that are never removed. Should we be ok with that?

it's an option to assert that listener list is empty instead, will revisit when we get it wired in.

**Pull request merged**